xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxMultiSource.cpp

```diff
@@ -0,0 +1,246 @@
+}
+
+void CDVDDemuxMultiSource::Abort()
+{
+  for (auto iter : m_pDemuxers)
+    iter->Abort();
+}
+
+void CDVDDemuxMultiSource::Dispose()
+{
+  while (!m_demuxerQueue.empty())
+  {
+    m_demuxerQueue.pop();
+  }
+
+  m_streamIdToDemuxerMap.clear();
+  m_DemuxerToInputStreamMap.clear();
+  m_pDemuxers.clear();
+  m_currentDemuxer = NULL;
+  m_pInput = NULL;
+
+}
+
+void CDVDDemuxMultiSource::Flush()
+{
+  for (auto iter : m_pDemuxers)
+    iter->Flush();
+}
+
+int CDVDDemuxMultiSource::GetNrOfStreams()
+{
+  int streamsCount = 0;
+  for (auto iter : m_pDemuxers)
+    streamsCount += iter->GetNrOfStreams();
+
+  return streamsCount;
+}
+
+CDemuxStream* CDVDDemuxMultiSource::GetStream(int iStreamId)
+{
+  auto iter = m_streamIdToDemuxerMap.find(iStreamId);
+  if (iter != m_streamIdToDemuxerMap.end())
+  {
+    return iter->second->GetStream(iStreamId);
+  }
+  else
+    return NULL;
+}
+
+void CDVDDemuxMultiSource::GetStreamCodecName(int iStreamId, std::string &strName)
+{
+  auto iter = m_streamIdToDemuxerMap.find(iStreamId);
+  if (iter != m_streamIdToDemuxerMap.end())
+  {
+    iter->second->GetStreamCodecName(iStreamId, strName);
+  }
+};
+
+int CDVDDemuxMultiSource::GetStreamLength()
+{
+  int length = 0;
+  for (auto iter : m_pDemuxers)
+  {
+    length = std::max(length, iter->GetStreamLength());
+  }
+
+  return length;
+}
+
+bool CDVDDemuxMultiSource::Open(CDVDInputStream* pInput)
+{
+  if (!pInput || !pInput->IsStreamType(DVDSTREAM_TYPE_MULTIFILES))
+    return false;
+
+  m_pInput = dynamic_cast<CDVDInputStreamMultiSource*>(pInput);
+
+  if (!m_pInput)
+    return false;
+
+  auto iter = m_pInput->m_pInputStreams.begin();
+  while (iter != m_pInput->m_pInputStreams.end())
+  {
+    DemuxPtr demuxer = DemuxPtr(CDVDFactoryDemuxer::CreateDemuxer(iter->get()));
+    if (!demuxer)
+    {
+      iter = m_pInput->m_pInputStreams.erase(iter);
+    }
+    else
+    {
+      unsigned int offset = m_streamIdToDemuxerMap.size();
+      demuxer->RenumberStreamIds(offset);
+      m_DemuxerStreamsOffset[demuxer] = offset;
+      if (!UpdateStreamMap(demuxer))
+      {
+        iter = m_pInput->m_pInputStreams.erase(iter);
+        m_DemuxerStreamsOffset.erase(demuxer);
+        continue;
+      }
+
+      m_DemuxerToInputStreamMap[demuxer] = *iter;
+      m_pDemuxers.push_back(demuxer);
+      m_demuxerQueue.push(std::make_pair((double)DVD_NOPTS_VALUE, demuxer));
+      ++iter;
+    }
+  }
+  return !m_pDemuxers.empty();
+}
+
+bool CDVDDemuxMultiSource::RebuildStreamMap()
+{
+  auto iter = std::find(m_pDemuxers.begin(), m_pDemuxers.end(), m_currentDemuxer);
+  m_DemuxerToInputStreamMap.erase(m_DemuxerToInputStreamMap.find(m_currentDemuxer));
+  m_pDemuxers.erase(iter);
+  m_currentDemuxer = NULL;
+
+  // TODO: 
+  // - rebuild m_DemuxerStreamsOffset
+  // - rebuild m_streamIdToDemuxerMap
+
+  return true;
+}
+
+void CDVDDemuxMultiSource::Reset()
+{
+  for (auto iter : m_pDemuxers)
+    iter->Reset();
+}
+
+DemuxPacket* CDVDDemuxMultiSource::Read()
```

Comment:
interestingly it's not, the managing of the streams, especially the ids, is the more problematic thing.
I'll add problem description to the pr post.